src/processes/soil/hydrology/soil_hydrology_rre.jl

```diff
@@ -0,0 +1,207 @@
+    grid = get_grid(model)
+    strat = get_stratigraphy(model)
+    bgc = get_biogeochemistry(model)
+    launch!(grid, :xyz, compute_hydraulic_conductivity!, state, grid, hydrology, strat, bgc)
+    return nothing
+end
+
+@kernel function compute_hydraulic_conductivity!(
+    state,
+    grid,
+    hydrology::SoilHydrology,
+    strat::AbstractStratigraphy,
+    bgc::AbstractSoilBiogeochemistry,
+)
+    i, j, k = @index(Global, NTuple)
+    soil = soil_composition((i, j, k), state, strat, hydrology, bgc)
+    state.hydraulic_conductivity[i, j, k] = hydraulic_conductivity(hydrology.hydraulic_properties, soil)
+    # Manually set hydraulic conductivity at the boundaries to the conductivity of the cell centers at the boundaries;
+    # note that this is an assumption that could potentially be relaxed in the future.
+    # TODO: maybe this should be done with boundary conditions?
+    field_grid = get_field_grid(grid) 
+    if k == 1
+        state.hydraulic_conductivity[i, j, 0] = state.hydraulic_conductivity[i, j, 1]
+    elseif k == field_grid.Nz
+        state.hydraulic_conductivity[i, j, field_grid.Nz+1] = state.hydraulic_conductivity[i, j, field_grid.Nz]
+    end
+end
+
+function compute_tendencies!(state, model, hydrology::SoilHydrology{NF, <:RichardsEq}) where {NF}
+    grid = get_grid(model)
+    strat = get_stratigraphy(model)
+    bgc = get_biogeochemistry(model)
+    launch!(grid, :xyz, compute_saturation_tendency!, state, grid, hydrology, strat, bgc)
+    return nothing
+end
+
+@kernel function compute_saturation_tendency!(
+    state,
+    grid,
+    hydrology::SoilHydrology,
+    strat::AbstractStratigraphy,
+    bgc::AbstractSoilBiogeochemistry,
+)
+    i, j, k = @index(Global, NTuple)
+    state.tendencies.saturation_water_ice[i, j, k] += saturation_tendency((i, j, k), grid, state, hydrology, strat, bgc)
+end
+
+@inline function saturation_tendency(
+    idx, grid, state,
+    hydrology::SoilHydrology{NF, <:RichardsEq},
+    strat::AbstractStratigraphy,
+    bgc::AbstractSoilBiogeochemistry
+) where {NF}
+    i, j, k = idx
+    # Operators require the underlying Oceananigans grid
+    field_grid = get_field_grid(grid)
+
+    # Compute divergence of water fluxes
+    # ∂θ∂t = ∇⋅K(θ)(∇ψ + 1) = ∇⋅K(θ)∇ψ
+    ∂θ∂t = ∂zᵃᵃᶜ(i, j, k, field_grid, darcy_flux, state)
+
+    # Get porosity
+    por = porosity(idx, state, hydrology, strat, bgc)
+
+    # Rescale volumetric flux to saturation flux
+    return -∂θ∂t / por
+end
+
+@inline function darcy_flux(i, j, k, grid, state)
+    # Get pressure head
+    ψ = state.pressure_head
+    # Get hydraulic conductivity
+    K = state.hydraulic_conductivity
+    # Darcy's law: q = -K ∂ψ/∂z
+    # TODO: also account for effect of temperature on matric potential
+    ∇ψ = ∂zᵃᵃᶠ(i, j, k, grid, ψ)
+    ## Take minimum of hydraulic conductivities in the direction of flow
+    Kₖ = (∇ψ < 0)*min(K[i, j, k-1], K[i, j, k]) +
```

Comment:
@glwagner I would appreciate your feedback on this approach. Maybe you have an idea for how this can be done better. Averaging using the interpolation operator is not viable for hydraulic conductivity because (1) it varies substantially (by order of magnitude) over space and (2) it is not physically justified. Taking the minimum of the hydraulic conductivities is a fairly common choice in hydrological modeling based on the physical intuition that the flow path of the water through the pore space is going to be limited by the most restrictive (least conductive) material along that path. Another option would be the harmonic mean.

However, I am using the minimum along the direction of flow here, which I am not sure about.

Reply:
I am not sure if I understand what you want, but you can define a function:

```julia
minᵃᵃᶠ(i, j, k, grid, a) = @inbounds min(a[i, j, k], a[i, j, k-1])
```

I don't agree that averaging is not physically justified. It is _mathematically_ justified (it is part of the staggered formulation). This is widely used in C-grid LES models, for example (not just Oceananigans --- all C-grid LES models for atmospheres and oceans). You can come up with another formulation for sure, but it is not correct to argue that averaging is not physical justified.

Reply:
It is indeed not physically justified because soil properties (e.g. saturation) are not continuous and hydraulic conductivities can vary by orders of magnitude between soil layers. Averaging K across cells is well known to produce wildly unrealistic soil water flow.

Take for example a very straightforward case of a frozen soil volume below a thawed soil volume. In the frozen volume, the conductivity is near zero since liquid water is immobile, but in the thawed volume, it is nonzero. Averaging would result in a nonzero flux into the frozen soil volume which is physically incorrect.

Reply:
Presumably you have a model for conductivity. In this model, you will interpolate the inputs of the model to cell interfaces, and then use the model to compute conductivity there.